app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,340 @@
+// when the counter is equal to 3, press_tap_dance_behavior will be called
+// to register N3's keycode being pressed
+static int on_tap_dance_binding_pressed(struct zmk_behavior_binding *binding,
+                                        struct zmk_behavior_binding_event event) {
+    // Code executes when the tap dance key is pressed.
+    const struct device *dev = device_get_binding(binding->behavior_dev);
+    const struct behavior_tap_dance_config *cfg = dev->config;
+    struct active_tap_dance *tap_dance;
+    tap_dance = find_tap_dance(event.position);
+    // No active tap dance found on keypress location, attempt to create a new one
+    if (tap_dance == NULL) {
+        if (new_tap_dance(event.position, cfg, &tap_dance) == -ENOMEM) {
+            LOG_ERR("Unable to create new tap dance. Insufficient space in active_tap_dances[].");
+            return ZMK_BEHAVIOR_OPAQUE;
+        }
+        // Tap dance successfully created. Reset timer.
+        LOG_DBG("%d created new tap dance", event.position);
+        reset_timer(tap_dance, event);
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    // An active tap dance was found at the keypress location. Stop the timer.
+    tap_dance->is_pressed = true;
+    LOG_DBG("%d tap dance re-pressed", event.position);
+    stop_timer(tap_dance);
+    // Increment the counter. Check if the counter has reached the maximum behavior number.
+    tap_dance->counter++;
+    if (tap_dance->counter >= cfg->behavior_count) {
+        // Counter is greater than or equal to the maximum number of behaviors.
+        // Resolve tap dance immediately.
+        press_tap_dance_behavior(tap_dance, event.timestamp);
+        tap_dance->tap_dance_decided = true;
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    // Counter is less than maximum number of behaviors, reset timer.
+    reset_timer(tap_dance, event);
+    return ZMK_BEHAVIOR_OPAQUE;
+}
+
+// on_tap_dance_binding_released
+
+// Determines what happens when a tap-dance binding is released.
+// Does not register a keycode release unless the tap-dance has
+// already been decided by its counter or the timer handler
+static int on_tap_dance_binding_released(struct zmk_behavior_binding *binding,
+                                         struct zmk_behavior_binding_event event) {
+    // Code executes when the tap dance key is released.
+    LOG_DBG("%d tap dance keybind released", event.position);
+    struct active_tap_dance *tap_dance = find_tap_dance(event.position);
+    if (tap_dance == NULL) {
+        LOG_ERR("ACTIVE TAP DANCE CLEARED TOO EARLY");
+        return ZMK_BEHAVIOR_OPAQUE;
+    }
+    tap_dance->is_pressed = false;
+    // An active tap dance has already been decided and is registering keypresses.
+    // Release tap dance on binding release.
+    if (tap_dance->tap_dance_decided) {
+        release_tap_dance_behavior(tap_dance, event.timestamp);
+        tap_dance->tap_dance_decided = false;
```

Comment:
`release_tap_dance_behavior` already calls `clear_tap_dance` so this is unnecessary